File: scripts/rank_companies.js
#!/usr/bin/env node

import { parseArgs as parseNodeArgs } from "node:util";
import { resolve } from "node:path";
import { readCsv, writeCsvAtomic } from "./lib/csv.js";
import { finiteNumber, isMain, projectDirectory, roundHalfEven } from "./lib/runtime.js";

export const EXCLUDED_SECTORS = new Set(["Financials", "Real Estate"]);
export const MIN_POSITIVE_FCF_YEARS = 4;
export const MIN_SECTOR_SAMPLE = 5;
export const MAX_PER_SECTOR = 2;

export const QUALITY_WEIGHTS = {
  operating_margin_score: 0.40,
  fcf_consistency_score: 0.35,
  revenue_growth_score: 0.25,
};

export const METRICS = {
  free_cash_flow_yield: [0.25, true, true],
  earnings_yield: [0.20, true, true],
  ev_to_operating_income: [0.15, false, true],
  operating_margin: [0.10, true, true],
  five_year_revenue_growth: [0.10, true, false],
  fcf_consistency: [0.10, true, false],
  net_debt_to_fcf: [0.10, false, false],
};

export const REQUIRED_FIELDS = new Set([
  "ticker", "cik", "name", "sector", "subsector", "valuation_price", "price_date",
  "free_cash_flow_yield", "earnings_yield", "ev_to_operating_income", "operating_margin",
  "five_year_revenue_growth", "positive_fcf_years", "historical_fcf_years",
  "net_debt_to_fcf", "five_year_median_fcf", "five_year_median_net_income",
  "calculation_status", "snapshot_source_url", "price_source_url",
]);

export const OUTPUT_FIELDS = [
  "rank", "ticker", "cik", "name", "sector", "subsector", "valuation_price", "price_date",
  "attractiveness_score", "quality_score", "quality_display_score", "quality_label",
  "free_cash_flow_yield", "free_cash_flow_yield_score", "earnings_yield",
  "earnings_yield_score", "ev_to_operating_income", "ev_to_operating_income_score",
  "operating_margin", "operating_margin_score", "five_year_revenue_growth",
  "revenue_growth_score", "positive_fcf_years", "historical_fcf_years",
  "fcf_consistency_score", "net_debt_to_fcf", "net_debt_score", "selection_reasons",
  "review_flags", "snapshot_source_url", "price_source_url",
];

export const SCORE_FIELDS = {
  free_cash_flow_yield: "free_cash_flow_yield_score",
  earnings_yield: "earnings_yield_score",
  ev_to_operating_income: "ev_to_operating_income_score",
  operating_margin: "operating_margin_score",
  five_year_revenue_growth: "revenue_growth_score",
  fcf_consistency: "fcf_consistency_score",
  net_debt_to_fcf: "net_debt_score",
};

export function parseArgs(argv = process.argv.slice(2)) {
  const root = projectDirectory(import.meta.url);
  const { values } = parseNodeArgs({
    args: argv,
    options: {
      input: { type: "string", default: `${root}/generated/data/latest_screen.csv` },
      output: { type: "string", default: `${root}/generated/data/latest_top10.csv` },
      limit: { type: "string", default: "10" },
      help: { type: "boolean", short: "h" },
    },
  });
  const limit = Number(values.limit);
  if (!Number.isInteger(limit)) throw new Error("--limit must be an integer.");
  return {
    input: resolve(values.input), output: resolve(values.output), limit, help: values.help ?? false,
  };
}

export function asNumber(value) {
  return finiteNumber(value);
}

export function loadScreen(path) {
  const { fields, rows } = readCsv(path);
  const missing = [...REQUIRED_FIELDS].filter((field) => !fields.includes(field)).sort();
  if (rows.length === 0 || missing.length) {
    throw new Error(`${path} is empty or missing required columns: ${missing.join(", ")}`);
  }
  return rows;
}

export function isEligible(row) {
  if (row.calculation_status !== "ok" || EXCLUDED_SECTORS.has(row.sector)) return false;
  for (const field of ["five_year_median_fcf", "five_year_median_net_income"]) {
    if ((asNumber(row[field]) ?? 0) <= 0) return false;
  }
  const positiveYears = asNumber(row.positive_fcf_years);
  const historicalYears = asNumber(row.historical_fcf_years);
  if (
    positiveYears === null || historicalYears === null
    || positiveYears < MIN_POSITIVE_FCF_YEARS || historicalYears <= 0
  ) return false;
  return Object.keys(METRICS)
    .filter((field) => field !== "fcf_consistency")
    .every((field) => asNumber(row[field]) !== null);
}

export function percentileScores(rows, metric, higherIsBetter) {
  const ordered = rows
    .map((row, index) => [index, Number(row[metric])])
    .sort((left, right) => left[1] - right[1]);
  if (ordered.length === 1) return new Map([[ordered[0][0], 50]]);

  const scores = new Map();
  let position = 0;
  while (position < ordered.length) {
    let end = position + 1;
    while (end < ordered.length && ordered[end][1] === ordered[position][1]) end += 1;
    const averageRank = (position + end - 1) / 2;
    let percentile = averageRank / (ordered.length - 1) * 100;
    if (!higherIsBetter) percentile = 100 - percentile;
    for (const [index] of ordered.slice(position, end)) scores.set(index, percentile);
    position = end;
  }
  return scores;
}

export function addMetricScores(rows) {
  const sectorGroups = new Map();
  rows.forEach((row, index) => {
    if (!sectorGroups.has(String(row.sector))) sectorGroups.set(String(row.sector), []);
    sectorGroups.get(String(row.sector)).push(index);
  });

  for (const [metric, [, higherIsBetter, sectorRelative]] of Object.entries(METRICS)) {
    const scoreField = SCORE_FIELDS[metric];
    if (metric === "fcf_consistency") {
      for (const row of rows) {
        row[scoreField] = Math.min(100, row.positive_fcf_years / row.historical_fcf_years * 100);
      }
      continue;
    }
    const globalScores = percentileScores(rows, metric, higherIsBetter);
    rows.forEach((row, index) => { row[scoreField] = globalScores.get(index); });
    if (sectorRelative) {
      for (const indexes of sectorGroups.values()) {
        if (indexes.length < MIN_SECTOR_SAMPLE) continue;
        const subset = indexes.map((index) => rows[index]);
        const localScores = percentileScores(subset, metric, higherIsBetter);
        indexes.forEach((originalIndex, localIndex) => {
          rows[originalIndex][scoreField] = localScores.get(localIndex);
        });
      }
    }
  }
}

export function formatPercent(value) {
  return `${(value * 100).toFixed(1)}%`;
}

export function selectionReasons(row) {
  const labels = {
    free_cash_flow_yield: `FCF yield ${formatPercent(row.free_cash_flow_yield)}`,
    earnings_yield: `earnings yield ${formatPercent(row.earnings_yield)}`,
    ev_to_operating_income: `EV/operating income ${row.ev_to_operating_income.toFixed(1)}x`,
    operating_margin: `operating margin ${formatPercent(row.operating_margin)}`,
    five_year_revenue_growth: `5-year revenue growth ${formatPercent(row.five_year_revenue_growth)}`,
    fcf_consistency: `positive FCF in ${Math.trunc(row.positive_fcf_years)}/${Math.trunc(row.historical_fcf_years)} years`,
    net_debt_to_fcf: `net debt/FCF ${row.net_debt_to_fcf.toFixed(1)}x`,
  };
  return Object.entries(METRICS)
    .map(([metric, [weight]]) => ({ contribution: row[SCORE_FIELDS[metric]] * weight, metric }))
    .sort((left, right) => (
      right.contribution - left.contribution
      || (left.metric < right.metric ? 1 : left.metric > right.metric ? -1 : 0)
    ))
    .slice(0, 3)
    .map(({ metric }) => labels[metric])
    .join("; ");
}

export function reviewFlags(row) {
  const flags = [];
  if (Number(row.free_cash_flow_yield) > 0.25) flags.push("extreme_fcf_yield_verify_inputs");
  if (Number(row.earnings_yield) > 0.25) flags.push("extreme_earnings_yield_verify_inputs");
  return flags.join(";");
}

export function calculateQualityScore(row) {
  return Object.entries(QUALITY_WEIGHTS)
    .reduce((total, [field, weight]) => total + Number(row[field]) * weight, 0);
}

export function qualityLabel(displayScore) {
  if (displayScore >= 8.5) return "Strong";
  if (displayScore >= 7.0) return "Good";
  if (displayScore >= 5.5) return "Fair";
  return "Weak";
}

export function rankCompanies(rows, limit = 10) {
  if (limit <= 0) throw new Error("--limit must be greater than zero.");
  const eligible = rows.filter(isEligible).map((source) => {
    const row = { ...source };
    for (const metric of Object.keys(METRICS)) {
      if (metric !== "fcf_consistency") row[metric] = Number(source[metric]);
    }
    row.positive_fcf_years = Number(source.positive_fcf_years);
    row.historical_fcf_years = Number(source.historical_fcf_years);
    return row;
  });
  if (eligible.length === 0) throw new Error("No companies satisfy the ranking eligibility rules.");

  addMetricScores(eligible);
  for (const row of eligible) {
    row.attractiveness_score = Object.entries(METRICS)
      .reduce((total, [metric, [weight]]) => total + row[SCORE_FIELDS[metric]] * weight, 0);
    row.quality_score = calculateQualityScore(row);
    row.quality_display_score = roundHalfEven(row.quality_score / 10, 1);
    row.quality_label = qualityLabel(row.quality_display_score);
  }
  eligible.sort((left, right) => (
    right.attractiveness_score - left.attractiveness_score
    || (String(left.ticker) < String(right.ticker) ? -1 : String(left.ticker) > String(right.ticker) ? 1 : 0)
  ));

  const deduplicated = [];
  const seenCiks = new Set();
  for (const row of eligible) {
    const cik = String(row.cik);
    if (seenCiks.has(cik)) continue;
    seenCiks.add(cik);
    deduplicated.push(row);
  }
  const selected = [];
  const sectors = new Map();
  for (const row of deduplicated) {
    const sector = String(row.sector);
    if ((sectors.get(sector) ?? 0) >= MAX_PER_SECTOR) continue;
    selected.push(row);
    sectors.set(sector, (sectors.get(sector) ?? 0) + 1);
    if (selected.length === limit) break;
  }
  if (selected.length < limit) {
    throw new Error(`Only ${selected.length} companies can be selected with the sector cap; requested ${limit}.`);
  }

  return selected.map((row, index) => {
    const rendered = Object.fromEntries(OUTPUT_FIELDS.map((field) => [field, String(row[field] ?? "")]));
    rendered.rank = String(index + 1);
    rendered.name = rendered.name.replace(/\|+$/, "").trim();
    rendered.attractiveness_score = row.attractiveness_score.toFixed(2);
    rendered.quality_score = row.quality_score.toFixed(2);
    rendered.quality_display_score = row.quality_display_score.toFixed(1);
    for (const scoreField of Object.values(SCORE_FIELDS)) rendered[scoreField] = row[scoreField].toFixed(2);
    rendered.positive_fcf_years = String(Math.trunc(row.positive_fcf_years));
    rendered.historical_fcf_years = String(Math.trunc(row.historical_fcf_years));
    rendered.selection_reasons = selectionReasons(row);
    rendered.review_flags = reviewFlags(row);
    return rendered;
  });
}

export async function main(argv = process.argv.slice(2)) {
  try {
    const args = parseArgs(argv);
    if (args.help) {
      console.log("Usage: node scripts/rank_companies.js [--input PATH] [--output PATH] [--limit N]");
      return 0;
    }
    const ranked = rankCompanies(loadScreen(args.input), args.limit);
    writeCsvAtomic(args.output, ranked, OUTPUT_FIELDS);
    console.log(`Wrote ${ranked.length} ranked companies to ${args.output}`);
    return 0;
  } catch (error) {
    console.error(`Error: ${error.message}`);
    return 1;
  }
}

if (isMain(import.meta.url)) process.exitCode = await main();

File: scripts/rank_companies.py
#!/usr/bin/env python3

"""Rank an auditable weekly research shortlist from latest_screen.csv.

The script deliberately calls the result a research shortlist, not a list of
stocks proven to be undervalued. It applies explicit eligibility rules, scores
seven valuation and quality signals, and limits sector concentration.

Usage:

    python3 scripts/rank_companies.py
    python3 scripts/rank_companies.py --limit 10 --output /tmp/top10.csv

Only the standard Python library is used.
"""

from __future__ import annotations

import argparse
import csv
import math
import os
import sys
import tempfile
from collections import Counter, defaultdict
from pathlib import Path
from typing import Sequence


EXCLUDED_SECTORS = {"Financials", "Real Estate"}
MIN_POSITIVE_FCF_YEARS = 4
MIN_SECTOR_SAMPLE = 5
MAX_PER_SECTOR = 2

QUALITY_WEIGHTS = {
    "operating_margin_score": 0.40,
    "fcf_consistency_score": 0.35,
    "revenue_growth_score": 0.25,
}

METRICS = {
    "free_cash_flow_yield": (0.25, True, True),
    "earnings_yield": (0.20, True, True),
    "ev_to_operating_income": (0.15, False, True),
    "operating_margin": (0.10, True, True),
    "five_year_revenue_growth": (0.10, True, False),
    "fcf_consistency": (0.10, True, False),
    "net_debt_to_fcf": (0.10, False, False),
}

REQUIRED_FIELDS = {
    "ticker",
    "cik",
    "name",
    "sector",
    "subsector",
    "valuation_price",
    "price_date",
    "free_cash_flow_yield",
    "earnings_yield",
    "ev_to_operating_income",
    "operating_margin",
    "five_year_revenue_growth",
    "positive_fcf_years",
    "historical_fcf_years",
    "net_debt_to_fcf",
    "five_year_median_fcf",
    "five_year_median_net_income",
    "calculation_status",
    "snapshot_source_url",
    "price_source_url",
}

OUTPUT_FIELDS = [
    "rank",
    "ticker",
    "cik",
    "name",
    "sector",
    "subsector",
    "valuation_price",
    "price_date",
    "attractiveness_score",
    "quality_score",
    "quality_display_score",
    "quality_label",
    "free_cash_flow_yield",
    "free_cash_flow_yield_score",
    "earnings_yield",
    "earnings_yield_score",
    "ev_to_operating_income",
    "ev_to_operating_income_score",
    "operating_margin",
    "operating_margin_score",
    "five_year_revenue_growth",
    "revenue_growth_score",
    "positive_fcf_years",
    "historical_fcf_years",
    "fcf_consistency_score",
    "net_debt_to_fcf",
    "net_debt_score",
    "selection_reasons",
    "review_flags",
    "snapshot_source_url",
    "price_source_url",
]

SCORE_FIELDS = {
    "free_cash_flow_yield": "free_cash_flow_yield_score",
    "earnings_yield": "earnings_yield_score",
    "ev_to_operating_income": "ev_to_operating_income_score",
    "operating_margin": "operating_margin_score",
    "five_year_revenue_growth": "revenue_growth_score",
    "fcf_consistency": "fcf_consistency_score",
    "net_debt_to_fcf": "net_debt_score",
}


def parse_args(argv: Sequence[str] | None = None) -> argparse.Namespace:
    project_dir = Path(__file__).resolve().parent.parent
    parser = argparse.ArgumentParser(
        description="Build a transparent weekly research shortlist."
    )
    parser.add_argument(
        "--input",
        type=Path,
        default=project_dir / "data" / "latest_screen.csv",
        help="Valuation screen CSV (default: ./data/latest_screen.csv).",
    )
    parser.add_argument(
        "--output",
        type=Path,
        default=project_dir / "data" / "latest_top10.csv",
        help="Output CSV (default: ./data/latest_top10.csv).",
    )
    parser.add_argument("--limit", type=int, default=10, help="Number of companies.")
    return parser.parse_args(argv)


def as_number(value: str | None) -> float | None:
    try:
        number = float((value or "").strip())
    except ValueError:
        return None
    return number if math.isfinite(number) else None


def load_screen(path: Path) -> list[dict[str, str]]:
    if not path.exists():
        raise RuntimeError(f"Input file does not exist: {path}")
    with path.open(newline="", encoding="utf-8-sig") as handle:
        reader = csv.DictReader(handle)
        rows = list(reader)
        fields = set(reader.fieldnames or [])
    missing = REQUIRED_FIELDS - fields
    if not rows or missing:
        raise RuntimeError(
            f"{path} is empty or missing required columns: {', '.join(sorted(missing))}"
        )
    return rows


def is_eligible(row: dict[str, str]) -> bool:
    if row.get("calculation_status") != "ok":
        return False
    if row.get("sector") in EXCLUDED_SECTORS:
        return False
    required_positive = (
        "five_year_median_fcf",
        "five_year_median_net_income",
    )
    if any((as_number(row.get(field)) or 0) <= 0 for field in required_positive):
        return False
    positive_years = as_number(row.get("positive_fcf_years"))
    historical_years = as_number(row.get("historical_fcf_years"))
    if (
        positive_years is None
        or historical_years is None
        or positive_years < MIN_POSITIVE_FCF_YEARS
        or historical_years <= 0
    ):
        return False
    score_inputs = [field for field in METRICS if field != "fcf_consistency"]
    return all(as_number(row.get(field)) is not None for field in score_inputs)


def percentile_scores(
    rows: list[dict[str, object]], metric: str, higher_is_better: bool
) -> dict[int, float]:
    """Return 0-100 percentile scores, assigning tied values an average rank."""
    ordered = sorted(
        ((index, float(row[metric])) for index, row in enumerate(rows)),
        key=lambda item: item[1],
    )
    if len(ordered) == 1:
        return {ordered[0][0]: 50.0}

    scores: dict[int, float] = {}
    position = 0
    while position < len(ordered):
        end = position + 1
        while end < len(ordered) and ordered[end][1] == ordered[position][1]:
            end += 1
        average_rank = (position + end - 1) / 2
        percentile = average_rank / (len(ordered) - 1) * 100
        if not higher_is_better:
            percentile = 100 - percentile
        for index, _ in ordered[position:end]:
            scores[index] = percentile
        position = end
    return scores


def add_metric_scores(rows: list[dict[str, object]]) -> None:
    sector_groups: dict[str, list[int]] = defaultdict(list)
    for index, row in enumerate(rows):
        sector_groups[str(row["sector"])].append(index)

    for metric, (_, higher_is_better, sector_relative) in METRICS.items():
        score_field = SCORE_FIELDS[metric]
        if metric == "fcf_consistency":
            for row in rows:
                row[score_field] = min(
                    100.0,
                    float(row["positive_fcf_years"])
                    / float(row["historical_fcf_years"])
                    * 100,
                )
            continue

        global_scores = percentile_scores(rows, metric, higher_is_better)
        for index, row in enumerate(rows):
            row[score_field] = global_scores[index]

        if sector_relative:
            for indexes in sector_groups.values():
                if len(indexes) < MIN_SECTOR_SAMPLE:
                    continue
                subset = [rows[index] for index in indexes]
                local_scores = percentile_scores(subset, metric, higher_is_better)
                for local_index, original_index in enumerate(indexes):
                    rows[original_index][score_field] = local_scores[local_index]


def format_percent(value: float) -> str:
    return f"{value * 100:.1f}%"


def selection_reasons(row: dict[str, object]) -> str:
    labels = {
        "free_cash_flow_yield": f"FCF yield {format_percent(float(row['free_cash_flow_yield']))}",
        "earnings_yield": f"earnings yield {format_percent(float(row['earnings_yield']))}",
        "ev_to_operating_income": f"EV/operating income {float(row['ev_to_operating_income']):.1f}x",
        "operating_margin": f"operating margin {format_percent(float(row['operating_margin']))}",
        "five_year_revenue_growth": f"5-year revenue growth {format_percent(float(row['five_year_revenue_growth']))}",
        "fcf_consistency": f"positive FCF in {int(float(row['positive_fcf_years']))}/{int(float(row['historical_fcf_years']))} years",
        "net_debt_to_fcf": f"net debt/FCF {float(row['net_debt_to_fcf']):.1f}x",
    }
    contributions = sorted(
        (
            (float(row[SCORE_FIELDS[metric]]) * weight, metric)
            for metric, (weight, _, _) in METRICS.items()
        ),
        reverse=True,
    )
    return "; ".join(labels[metric] for _, metric in contributions[:3])


def review_flags(row: dict[str, object]) -> str:
    flags: list[str] = []
    if float(row["free_cash_flow_yield"]) > 0.25:
        flags.append("extreme_fcf_yield_verify_inputs")
    if float(row["earnings_yield"]) > 0.25:
        flags.append("extreme_earnings_yield_verify_inputs")
    return ";".join(flags)


def calculate_quality_score(row: dict[str, object]) -> float:
    return sum(float(row[field]) * weight for field, weight in QUALITY_WEIGHTS.items())


def quality_label(display_score: float) -> str:
    if display_score >= 8.5:
        return "Strong"
    if display_score >= 7.0:
        return "Good"
    if display_score >= 5.5:
        return "Fair"
    return "Weak"


def rank_companies(rows: list[dict[str, str]], limit: int = 10) -> list[dict[str, str]]:
    if limit <= 0:
        raise RuntimeError("--limit must be greater than zero.")

    eligible: list[dict[str, object]] = []
    for source in rows:
        if not is_eligible(source):
            continue
        row: dict[str, object] = dict(source)
        for metric in METRICS:
            if metric == "fcf_consistency":
                continue
            row[metric] = float(source[metric])
        row["positive_fcf_years"] = float(source["positive_fcf_years"])
        row["historical_fcf_years"] = float(source["historical_fcf_years"])
        eligible.append(row)

    if not eligible:
        raise RuntimeError("No companies satisfy the ranking eligibility rules.")

    add_metric_scores(eligible)
    for row in eligible:
        row["attractiveness_score"] = sum(
            float(row[SCORE_FIELDS[metric]]) * weight
            for metric, (weight, _, _) in METRICS.items()
        )
        row["quality_score"] = calculate_quality_score(row)
        row["quality_display_score"] = round(float(row["quality_score"]) / 10, 1)
        row["quality_label"] = quality_label(float(row["quality_display_score"]))

    ordered = sorted(
        eligible,
        key=lambda row: (-float(row["attractiveness_score"]), str(row["ticker"])),
    )
    deduplicated: list[dict[str, object]] = []
    seen_ciks: set[str] = set()
    for row in ordered:
        cik = str(row["cik"])
        if cik in seen_ciks:
            continue
        seen_ciks.add(cik)
        deduplicated.append(row)

    selected: list[dict[str, object]] = []
    sectors: Counter[str] = Counter()
    for row in deduplicated:
        sector = str(row["sector"])
        if sectors[sector] >= MAX_PER_SECTOR:
            continue
        selected.append(row)
        sectors[sector] += 1
        if len(selected) == limit:
            break

    if len(selected) < limit:
        raise RuntimeError(
            f"Only {len(selected)} companies can be selected with the sector cap; "
            f"requested {limit}."
        )

    output: list[dict[str, str]] = []
    for rank, row in enumerate(selected, start=1):
        rendered = {field: str(row.get(field, "")) for field in OUTPUT_FIELDS}
        rendered["rank"] = str(rank)
        rendered["name"] = rendered["name"].rstrip("|").strip()
        rendered["attractiveness_score"] = f"{float(row['attractiveness_score']):.2f}"
        rendered["quality_score"] = f"{float(row['quality_score']):.2f}"
        rendered["quality_display_score"] = f"{float(row['quality_display_score']):.1f}"
        for score_field in SCORE_FIELDS.values():
            rendered[score_field] = f"{float(row[score_field]):.2f}"
        rendered["positive_fcf_years"] = str(int(float(row["positive_fcf_years"])))
        rendered["historical_fcf_years"] = str(int(float(row["historical_fcf_years"])))
        rendered["selection_reasons"] = selection_reasons(row)
        rendered["review_flags"] = review_flags(row)
        output.append(rendered)
    return output


def write_atomic(path: Path, rows: list[dict[str, str]]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    temporary_path: Path | None = None
    try:
        with tempfile.NamedTemporaryFile(
            "w",
            newline="",
            encoding="utf-8",
            dir=path.parent,
            prefix=f".{path.name}.",
            suffix=".tmp",
            delete=False,
        ) as handle:
            temporary_path = Path(handle.name)
            writer = csv.DictWriter(handle, fieldnames=OUTPUT_FIELDS, lineterminator="\n")
            writer.writeheader()
            writer.writerows(rows)
            handle.flush()
            os.fsync(handle.fileno())
        os.replace(temporary_path, path)
    finally:
        if temporary_path and temporary_path.exists():
            temporary_path.unlink()


def main(argv: Sequence[str] | None = None) -> int:
    args = parse_args(argv)
    try:
        source = load_screen(args.input)
        ranked = rank_companies(source, args.limit)
        write_atomic(args.output, ranked)
    except (OSError, RuntimeError) as error:
        print(f"Error: {error}", file=sys.stderr)
        return 1
    print(f"Wrote {len(ranked)} ranked companies to {args.output}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
